client/protocols/analytics.py

```diff
@@ -89,5 +96,87 @@ def replaced(self, contents):
             return False
         return True
 
+    @classmethod
+    def read_history(cls):
+        history = {'questions': {}, 'all_attempts': 0}
+        try:
+            with open(cls.ANALYTICS_FILE, 'rb') as fp:
+                history = pickle.load(fp)
+            log.info('Loaded %d history from %s',
+                     len(history), cls.ANALYTICS_FILE)
+        except (IOError, EOFError) as e:
+            log.info('Error reading from ' + cls.ANALYTICS_FILE + \
+                     ', assume no history')
+        return history
+
+    def log_run(self, messages):
+        """Record this run of the autograder to a local file.
+        """
+        history = self.read_history()
+        history['all_attempts'] += 1
+        analytics = messages['analytics']
+        questions = analytics.get('question', [])
+        grading = 'grading' in messages and messages['grading']
+
+        # Attempt to figure out what question is being worked on
```

Comment:
Please elaborate more on what "worked on" means, e.g. "the student is currently solving/implementing"